lib/scholar/decomposition/truncated_svd.ex

```diff
@@ -0,0 +1,225 @@
+        each component.
+
+    * `:explained_variance_ratio` - tensor of shape `{n_components}`
+        Percentage of variance explained by each of the selected components.
+
+    * `:singular_values` -  ndarray of shape `{n_components}`
+        The singular values corresponding to each of the selected components.
+
+  ## Examples
+
+      iex> x = Nx.tensor([[0, 0], [1, 0], [1, 1], [3, 3], [4, 4.5]])
+      iex> tsvd = Scholar.Decomposition.TruncatedSVD.fit(x, n_components: 2)
+      iex> tsvd.components
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [0.6871105432510376, 0.7265529036521912],
+          [0.7265529036521912, -0.6871105432510376]
+        ]
+      >
+      iex> tsvd.singular_values
+      #Nx.Tensor<
+        f32[2]
+        [7.528080940246582, 0.7601959705352783]
+      >
+  """
+
+  deftransform fit(x, opts \\ []) do
+    fit_n(x, NimbleOptions.validate!(opts, @tsvd_schema))
+  end
+
+  @doc """
+  Fit model to X and perform dimensionality reduction on X.
+
+  ## Options
+
+  #{NimbleOptions.docs(@tsvd_schema)}
+
+  ## Return Values
+
+    X_new tensor of shape `{n_samples, n_components}` - reduced version of X. 
+
+  ## Examples
+
+      iex> x = Nx.tensor([[0, 0], [1, 0], [1, 1], [3, 3], [4, 4.5]])
+      iex> Scholar.Decomposition.TruncatedSVD.fit_transform(x, n_components: 2)
+      #Nx.Tensor<
+        f32[5][2]
+        [
+          [0.0, 0.0],
+          [0.6871105432510376, 0.7265529036521912],
+          [1.413663387298584, 0.039442360401153564],
+          [4.240990161895752, 0.1183270812034607],
+          [6.017930030822754, -0.18578583002090454]
+        ]
+      >
+  """
+
+  deftransform fit_transform(x, opts \\ []) do
+    fit_transform_n(x, NimbleOptions.validate!(opts, @tsvd_schema))
+  end
+
+  defnp fit_n(x, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError, "expected x to have rank equal to: 2, got: #{inspect(Nx.rank(x))}"
+    end
+
+    n_components = opts[:n_components]
+    {_n_samples, n_features} = Nx.shape(x)
+
+    cond do
+      n_components > n_features ->
+        raise ArgumentError,
+              """
+              n_components must be less than or equal to \
+              n_features = #{n_features}, got #{n_components}
+              """
+
+      true ->
+        nil
+    end
+
+    {u, sigma, vt} = randomized_svd(x, opts)
+    {_u, vt} = Scholar.Decomposition.PCA.flip_svd(u, vt)
+
+    x_transformed = Nx.dot(x, Nx.transpose(vt))
+    explained_variance = Nx.variance(x_transformed, axes: [0])
+    full_variance = Nx.variance(x, axes: [0]) |> Nx.sum()
+    explained_variance_ratio = explained_variance / full_variance
+
+    %__MODULE__{
+      components: vt,
+      explained_variance: explained_variance,
+      explained_variance_ratio: explained_variance_ratio,
+      singular_values: sigma
+    }
+  end
+
+  defnp fit_transform_n(x, opts) do
+    module = fit_n(x, opts)
+    Nx.dot(x, Nx.transpose(module.components))
```

Comment:
Use here `Nx.dot/4` which automatically transposes particular tensors, depending on axes you provided